### PR TITLE
Add benchmark support for vector radial search

### PR DESCRIPTION
**Pull request by junqiu-lei**
### Description
Since OpenSearch version 2.14, we've introduced vector [radial search](https://opensearch.org/docs/latest/search-plugins/knn/radial-search-knn/) in k-NN plugin, this PR will support run benchmark with radial search api.

Raised another PR https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/309 in opensearch-benchmark-workloads to have the change accordinately.

### Testing
- [x] New functionality includes testing

Example local run workloads and results:
```
{
    "target_index_name": "target_index",
    "target_field_name": "target_field",
    "target_index_body": "indices/faiss-index.json",
    "target_index_primary_shards": 3,
    "target_index_dimension": 768,
    "target_index_space_type": "innerproduct",
    
    "target_index_bulk_size": 100,
    "target_index_bulk_index_data_set_format": "hdf5",
    "target_index_bulk_index_data_set_path": "/Users/junqiu/dataset/documents-1m-threshold-innerproduct-160.hdf5",
    "target_index_bulk_indexing_clients": 10,
    
    "target_index_max_num_segments": 1,
    "target_index_force_merge_timeout": 600.0,
    "hnsw_ef_search": 256,
    "hnsw_ef_construction": 256,
    "target_index_num_vectors": 1000000,
    "query_max_distance": -160.0,
    "query_body": {
         "docvalue_fields" : ["_id"],
         "stored_fields" : "_none_"
    },

    "query_data_set_format": "hdf5",
    "query_data_set_path":"/Users/junqiu/dataset/documents-1m-threshold-innerproduct-160.hdf5",
    "query_count": 100
}
```

```
|---------------------------------------------------------------:|-------------:|------------:|-------:|
|                     Cumulative indexing time of primary shards |              |     83.6451 |    min |
|             Min cumulative indexing time across primary shards |              | 0.000416667 |    min |
|          Median cumulative indexing time across primary shards |              |   0.0165833 |    min |
|             Max cumulative indexing time across primary shards |              |     28.2067 |    min |
|            Cumulative indexing throttle time of primary shards |              |           0 |    min |
|    Min cumulative indexing throttle time across primary shards |              |           0 |    min |
| Median cumulative indexing throttle time across primary shards |              |           0 |    min |
|    Max cumulative indexing throttle time across primary shards |              |           0 |    min |
|                        Cumulative merge time of primary shards |              |     93.4468 |    min |
|                       Cumulative merge count of primary shards |              |          94 |        |
|                Min cumulative merge time across primary shards |              |           0 |    min |
|             Median cumulative merge time across primary shards |              |           0 |    min |
|                Max cumulative merge time across primary shards |              |     32.9337 |    min |
|               Cumulative merge throttle time of primary shards |              |     3.82742 |    min |
|       Min cumulative merge throttle time across primary shards |              |           0 |    min |
|    Median cumulative merge throttle time across primary shards |              |           0 |    min |
|       Max cumulative merge throttle time across primary shards |              |     1.39933 |    min |
|                      Cumulative refresh time of primary shards |              |     4.67383 |    min |
|                     Cumulative refresh count of primary shards |              |         211 |        |
|              Min cumulative refresh time across primary shards |              | 0.000866667 |    min |
|           Median cumulative refresh time across primary shards |              |      0.0031 |    min |
|              Max cumulative refresh time across primary shards |              |     1.66205 |    min |
|                        Cumulative flush time of primary shards |              |     3.79628 |    min |
|                       Cumulative flush count of primary shards |              |          42 |        |
|                Min cumulative flush time across primary shards |              |           0 |    min |
|             Median cumulative flush time across primary shards |              |  0.00796667 |    min |
|                Max cumulative flush time across primary shards |              |     1.29665 |    min |
|                                        Total Young Gen GC time |              |           0 |      s |
|                                       Total Young Gen GC count |              |           0 |        |
|                                          Total Old Gen GC time |              |           0 |      s |
|                                         Total Old Gen GC count |              |           0 |        |
|                                                     Store size |              |     17.0813 |     GB |
|                                                  Translog size |              | 3.58559e-07 |     GB |
|                                         Heap used for segments |              |           0 |     MB |
|                                       Heap used for doc values |              |           0 |     MB |
|                                            Heap used for terms |              |           0 |     MB |
|                                            Heap used for norms |              |           0 |     MB |
|                                           Heap used for points |              |           0 |     MB |
|                                    Heap used for stored fields |              |           0 |     MB |
|                                                  Segment count |              |          23 |        |
|                                                 Min Throughput | prod-queries |       66.96 |  ops/s |
|                                                Mean Throughput | prod-queries |       66.96 |  ops/s |
|                                              Median Throughput | prod-queries |       66.96 |  ops/s |
|                                                 Max Throughput | prod-queries |       66.96 |  ops/s |
|                                        50th percentile latency | prod-queries |     5.82227 |     ms |
|                                        90th percentile latency | prod-queries |     13.9878 |     ms |
|                                        99th percentile latency | prod-queries |     63.4047 |     ms |
|                                       100th percentile latency | prod-queries |     90.4701 |     ms |
|                                   50th percentile service time | prod-queries |     5.82227 |     ms |
|                                   90th percentile service time | prod-queries |     13.9878 |     ms |
|                                   99th percentile service time | prod-queries |     63.4047 |     ms |
|                                  100th percentile service time | prod-queries |     90.4701 |     ms |
|                                                     error rate | prod-queries |           0 |      % |
```
---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by gkamat — CHANGES_REQUESTED** [review]
Hi @junqiu-lei, a large portion of the diffs pertain to formatting changes unrelated to the changed/new functionality.  Can you submit only the latter in this PR?

We want to be cautious about changing formatting guidelines.  Suggestions are welcome, but they need to be discussed separately and applied globally.   Thanks for understanding.

**Comment by junqiu-lei** [comment]
> Hi @junqiu-lei, a large portion of the diffs pertain to formatting changes unrelated to the changed/new functionality. Can you submit only the latter in this PR?
> 
> We want to be cautious about changing formatting guidelines. Suggestions are welcome, but they need to be discussed separately and applied globally. Thanks for understanding.

@gkamat Thanks the feedback, yes, I just updated PR.

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1083]: we can't remove top_k. It was added to support top k where k < len(neighbors[0])

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1100]: why do we need this?

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1081]: i believe we need validation to support either k or min/max

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1135]: move it as constant

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py:1082]: Updated

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py]: Updated

File: osbenchmark/worker_coordinator/runner.py
Comment (junqiu-lei) [reply osbenchmark/worker_coordinator/runner.py:1083]: Created another individual function for calculating radial search recall

**Review by IanHoang — APPROVED** [review]

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1155]: why do we need to check for k here?

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1156]: if you want to check whether max_distance is in params, use "key" in params, else it will be evaluated to false if value is zero

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1195]: how about raise error if k and max/min is provided by params?

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1049]: nit: DEFAULT_RADIAL_SEARCH_QUERY_RESULT_SIZE

Comment (VijayanB) [reply osbenchmark/workload/params.py:1049]: is there a way to override this by params?

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [reply osbenchmark/worker_coordinator/runner.py:1195]: i see validation added there. Can you explicitly check for params like k at line 1201 instead of else. We don't want catch all clause if it is not k.

File: tests/worker_coordinator/runner_test.py
Comment (VijayanB) [inline tests/worker_coordinator/runner_test.py:2808]: can you add test for if both k and min/max is provided?

**Comment by VijayanB** [comment]
@junqiu-lei Overall looks good. Added minor comments. You might also need follow up PR once #581 is merged

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py:1049]: No, as of now I don't see the demand of setting it as parameter

File: osbenchmark/worker_coordinator/runner.py
Comment (junqiu-lei) [reply osbenchmark/worker_coordinator/runner.py:1156]: @VijayanB I don't get your point here.

File: tests/worker_coordinator/runner_test.py
Comment (junqiu-lei) [reply tests/worker_coordinator/runner_test.py:2808]: Added tests

File: osbenchmark/worker_coordinator/runner.py
Comment (junqiu-lei) [reply osbenchmark/worker_coordinator/runner.py]: updated

File: osbenchmark/worker_coordinator/runner.py
Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1199]: can you move  logging.getLogger(__name__) as variable like self.logger

Comment (VijayanB) [inline osbenchmark/worker_coordinator/runner.py:1199]: Why do we need to log this? This will flood the log file

Comment (junqiu-lei) [reply osbenchmark/worker_coordinator/runner.py]: Removed

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1133]:
```suggestion
        if self.query_type == self.KNN_QUERY_TYPE:
            return Context.NEIGHBORS
        if self.query_type == self.MIN_SCORE_QUERY_TYPE:
            return Context.MIN_SCORE_NEIGHBORS
        if self.query_type == self.MAX_DISTANCE_QUERY_TYPE:
            return Context.MAX_DISTANCE_NEIGHBORS
        raise exceptions.InvalidSyntax("Unknown query type [%s]" % self.query_type)
           

```

Comment (junqiu-lei) [reply osbenchmark/workload/params.py]: Updated

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1067]: remove this logging as well

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1131]: you don't need else here

File: osbenchmark/workload/params.py
Comment (VijayanB) [inline osbenchmark/workload/params.py:1214]: can we move k to top?

**Review by VijayanB — COMMENTED** [review]
LGTM overall. Just added few minor comments.

**Review by VijayanB — APPROVED** [review]

File: osbenchmark/workload/params.py
Comment (gkamat) [inline osbenchmark/workload/params.py:1129]: Using `elif` will be more appropriate.

File: osbenchmark/workload/params.py
Comment (gkamat) [inline osbenchmark/workload/params.py:1139]: Breaking it by type will be clearer, like in `_get_query_neighbors()` above.

File: osbenchmark/workload/params.py
Comment (gkamat) [inline osbenchmark/workload/params.py:1181]:
```
true_neighbors = list(map(str, neighbor[:self.k] if self.k else neigbor)
```

File: osbenchmark/workload/params.py
Comment (gkamat) [inline osbenchmark/workload/params.py:1204]: Instead of using `query.update()` here:
```
query_update["k"] = self.k
```

File: osbenchmark/worker_coordinator/runner.py
Comment (junqiu-lei) [reply osbenchmark/worker_coordinator/runner.py:1319]: ack

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py:1106]: ack

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py:1129]: Ummm, I used to use `elif`, @VijayanB and I agreed to updated to current condition check for cleaner read. It has validation step before this

File: osbenchmark/workload/params.py
Comment (junqiu-lei) [reply osbenchmark/workload/params.py:1204]: I'm following the same pattern used in this function.

**Review by VijayanB — APPROVED** [review]

File: osbenchmark/workload/params.py
Comment (gkamat) [reply osbenchmark/workload/params.py:1128]: The effect is the same in this case, but using `elif` makes it more readable, especially since the cases are mutually exclusive.  I'll leave it to you.

File: osbenchmark/workload/params.py
Comment (gkamat) [reply osbenchmark/workload/params.py:1202]: Again, this is for better readability, especially if there is only a single key being updated.  I won't insist, though.

**Review by gkamat — APPROVED** [review]

**Pull request merged**
